spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,505 @@
+
+    ExecutorService rewriteService = Executors.newFixedThreadPool(maxConcurrentFileGroupRewrites,
+        new ThreadFactoryBuilder().setNameFormat("Rewrite-Service-%d").build());
+
+    ExecutorService committerService = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setNameFormat("Committer-Service").build());
+
+    int groupsPerCommit = IntMath.divide(ctx.totalGroupCount(), maxCommits, RoundingMode.CEILING);
+
+    AtomicBoolean stillRewriting = new AtomicBoolean(true);
+    ConcurrentHashMap<String, Pair<FileGroupInfo, FileGroupRewriteResult>> results = new ConcurrentHashMap<>();
+    ConcurrentLinkedQueue<String> completedRewriteIds = new ConcurrentLinkedQueue<>();
+
+    // Partial progress commit service
+    committerService.execute(() -> {
+      while (stillRewriting.get() || completedRewriteIds.size() > 0) {
+        Thread.yield();
+        // Either we have a full commit group, or we have completed writing and need to commit what is left over
+        if (completedRewriteIds.size() >= groupsPerCommit ||
+            (!stillRewriting.get() && completedRewriteIds.size() > 0)) {
+
+          Set<String> batch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+          for (int i = 0; i < groupsPerCommit && !completedRewriteIds.isEmpty(); i++) {
+            batch.add(completedRewriteIds.poll());
+          }
+
+          try {
+            commitOrClean(batch);
+          } catch (Exception e) {
+            batch.forEach(results::remove);
+            LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+          }
+        }
+      }
+    });
+
+    // Start rewrite tasks
+    Tasks.foreach(groupStream.iterator())
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure((fileGroupForRewrite, exception) -> {
+          LOG.error("Failure during rewrite process for group {}", fileGroupForRewrite.info, exception);
+          abortFileGroup(fileGroupForRewrite.groupID()); })
+        .run(infoListPair -> rewriteFiles(ctx, infoListPair, completedRewriteIds, results));
+
+    stillRewriting.set(false);
+    committerService.shutdown();
+
+    try {
+      committerService.awaitTermination(10, TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Cannot complete commit for rewrite, commit service interrupted", e);
+    }
+
+    return new Result(results.values().stream().collect(Collectors.toMap(Pair::first, Pair::second)));
+  }
+
+  private Stream<FileGroupForRewrite> toGroupStream(RewriteExecutionContext ctx,
+      Map<StructLike, List<List<FileScanTask>>> fileGroupsByPartition) {
+
+    // Todo Add intelligence to the order in which we do rewrites instead of just using partition order
+    return fileGroupsByPartition.entrySet().stream()
+        .flatMap(
+            e -> e.getValue().stream().map(tasks -> {
+              int myJobIndex = ctx.currentGlobalIndex();
```

Comment:
nit: `currentGlobalIndex`, `currentPartitionIndex`?

Reply:
switched to those - current so the "new FileGroup" fits on one line